### PR TITLE
Backport DataTables using merge commit

### PR DESCRIPTION
**Pull request by nalimilan**
This is equivalent to https://github.com/JuliaStats/DataFrames.jl/pull/1214 but using `git merge` rather than `git rebase`. The merge commit is quite small (it can be seen locally using `git show` from the head of the branch). DataTables history has been preserved, except for the DataFrames ->DataTables rename which has been erased using [this script](https://gist.github.com/nalimilan/9c7c861dfafbc3601487bd03694a0931).

### CODE REVIEW

**Comment by nalimilan** [comment]
Of course this cannot be merged via the GitHub interface since that's not fast-forward. It will have to be done manually.

**Comment by nalimilan** [comment]
(Even if the colors don't indicate it, tests are green on Travis with Julia 0.6.)

**Comment by rofinn** [comment]
Diffing `rf/datatables` and `nl/datatables-backport` wasn't very helpful for reviewing, but a few things I did notice were:

1. A few remaining incorrect names (e.g., `dt` -> `df`, JuliaData -> JuliaStats).
2. Documentation that needs to be removed or fixed (e.g., the formulas documentation should be removed)

**Comment by nalimilan** [comment]
Good catch! New version should fix these issues.

File: docs/src/lib/utilities.md [hunk 1]
```diff
@@ -9,14 +8,13 @@ CurrentModule = DataFrames
 Pages = ["utilities.md"]
 ```
 
-...
-
 ```@docs
 eltypes
 head
 completecases
-completecases!
```
Comment (ararslan) [inline docs/src/lib/utilities.md]: Shouldn't we be removing `completecases` here in addition to `completecases!`?

File: docs/src/man/reshaping_and_pivoting.md [hunk 2]
```diff
@@ -3,8 +3,9 @@
 Reshape data from wide to long format using the `stack` function:
 
 ```julia
-using DataFrames, RDatasets
-iris = dataset("datasets", "iris")
+using DataFrames
+using CSV
+iris = CSV.read(joinpath(Pkg.dir("DataFrames"), "test/data/iris.csv"), DataFrame)
```
Comment (ararslan) [inline docs/src/man/reshaping_and_pivoting.md]: The slashes in the path here kind of defeat the purpose of using `joinpath` 😉

File: src/DataFrames.jl [hunk 3]
```diff
@@ -25,38 +19,23 @@ import Base: ==, |>
...
-       ContrastsCoding,
 
        aggregate,
        by,
-       coefnames,
+       categorical!,
        colwise,
        combine,
        completecases,
-       completecases!,
```
Comment (ararslan) [inline src/DataFrames.jl]: Same comment regarding `completecases`

File: src/abstractdataframe/abstractdataframe.jl [hunk 4]
```diff
@@ -229,23 +226,17 @@ Base.ndims(::AbstractDataFrame) = 2
...
+    DataFrame(Any[similar_nullable(x, dims) for x in columns(df)], copy(index(df)))
 
 ##############################################################################
 ##
 ## Equality
 ##
 ##############################################################################
 
+# Imported in DataFrames.jl for compatibility across Julia 0.4 and 0.5
+Base.:(==)(df1::AbstractDataFrame, df2::AbstractDataFrame) = isequal(df1, df2)
```
Comment (ararslan) [inline src/abstractdataframe/abstractdataframe.jl]: Didn't we change this recently in DataTables? Having `==` call `isequal` defeats the purpose of having them be separate functions. If this is only for 0.4 and 0.5 compatibility, we don't support those anymore, so this could be removed.

File: src/abstractdataframe/abstractdataframe.jl [hunk 5]
```diff
@@ -405,14 +378,43 @@ function StatsBase.describe(io, df::AbstractDataFrame)
     end
 end
 
+function StatsBase.describe{T}(io::IO, X::AbstractVector{Union{T, Null}})
```
Comment (ararslan) [inline src/abstractdataframe/abstractdataframe.jl]: This syntax is deprecated, should use `where`

File: src/abstractdataframe/io.jl [hunk 6]
```diff
@@ -5,7 +5,7 @@
 ##############################################################################
 
 function escapedprint(io::IO, x::Any, escapes::AbstractString)
-    print(io, x)
+    ourshowcompact(io, x)
```
Comment (ararslan) [inline src/abstractdataframe/io.jl]: Where are we defining this? It doesn't appear to be defined in this file.

File: src/abstractdataframe/join.jl [hunk 7]
```diff
@@ -2,136 +2,191 @@
...
-
-        if rc > 0 && lc > 0
-            tcount += lc * rc
-        elseif rc > 0
-            rcount += rc
-        else
-            lcount += lc
-        end
+# Like similar, but returns a array that can have nulls and is initialized with nulls
+similar_nullable{T}(dv::AbstractArray{T}, dims::Union{Int, Tuple{Vararg{Int}}}) =
```
Comment (ararslan) [inline src/abstractdataframe/join.jl]: `where`

File: src/abstractdataframe/join.jl [hunk 8]
```diff
@@ -2,136 +2,191 @@
...
-                    left_indexer[offset + k] = left_pos + j
-                    right_indexer[offset + k] = right_pos + k
-                end
-            end
-            tposition += lc * rc
-        end
-        left_pos += lc
-        right_pos += rc
-    end
+DataFrameJoiner{DF1<:AbstractDataFrame, DF2<:AbstractDataFrame}(dfl::DF1, dfr::DF2, on::Union{Symbol,Vector{Symbol}}) =
```
Comment (ararslan) [inline src/abstractdataframe/join.jl]: `where`

File: src/abstractdataframe/join.jl [hunk 9]
```diff
@@ -2,136 +2,191 @@
...
-    #     might be faster.
-    refs = zeros(R, nrow(df))
-    poolref = Dict{AbstractDataFrame, Int}()
-    pool = Vector{UInt64}(0)
-    j = 1
-    for i = 1:nrow(df)
-        val = df[i,:]
-        if haskey(poolref, val)
-            refs[i] = poolref[val]
+function fillcolumn!{T1, T2}(dfcol::AbstractVector{T1}, refcol::AbstractVector{T2},
```
Comment (ararslan) [inline src/abstractdataframe/join.jl]: `where`

File: src/abstractdataframe/sort.jl [hunk 10]
```diff
@@ -55,69 +55,36 @@ ordering(col::ColumnIndex, lt::Function, by::Function, rev::Bool, order::Orderin
 #         the permutation induced by this ordering is used to
 #         sort the original (presumably larger) DataFrame
 
-type DFPerm{O<:Union{Ordering, AbstractVector}, DF<:AbstractDataFrame} <: Ordering
+immutable DFPerm{O<:Union{Ordering, AbstractVector}, DF<:AbstractDataFrame} <: Ordering
```
Comment (ararslan) [inline src/abstractdataframe/sort.jl]: `struct`

File: src/dataframe/dataframe.jl [hunk 11]
```diff
@@ -689,7 +663,7 @@ function deleterows!(df::DataFrame, ind::AbstractVector{Int})
     idf = 1
     iind = 1
     ikeep = 1
-    keep = Vector{Int}(n - length(ind2))
+    keep = Array{Int}(n-length(ind2))
```
Comment (ararslan) [inline src/dataframe/dataframe.jl]: ?

File: src/other/utils.jl [hunk 12]
```diff
@@ -96,95 +95,54 @@ end
 #'
 #' DataFrames.gennames(10)
 function gennames(n::Integer)
-    res = Vector{Symbol}(n)
+    res = Array{Symbol}(n)
```
Comment (ararslan) [inline src/other/utils.jl]: Another seemingly unnecessary change from `Vector` to `Array`

File: src/other/utils.jl [hunk 13]
```diff
@@ -96,95 +95,54 @@ end
...
 #'
-#' @returns count::Int The number of missing values in `a`.
+#' @returns count::Int The number of null values in `a`.
 #'
 #' @examples
 #'
-#' DataFrames.countna(@data([1, 2, 3]))
-countna(da::DataArray) = sum(da.na)
+#' DataFrames.countnull([1, 2, 3])
+function countnull(a::AbstractArray)
```
Comment (ararslan) [inline src/other/utils.jl]: Seems like this should either be defined in Nulls or deprecated in favor of `count(isnull, a)`

File: src/subdataframe/subdataframe.jl [hunk 14]
```diff
@@ -5,6 +5,42 @@
 ##
 ##############################################################################
 
+if VERSION >= v"0.6.0-dev.2643"
```
Comment (ararslan) [inline src/subdataframe/subdataframe.jl]: Could simplify this to remove the conditional

File: test/data.jl [hunk 15]
```diff
@@ -42,18 +34,24 @@ module TestData
     # lots more to do
 
     #test_group("assign")
-    df6[3] = @data(["un", "deux", "troix", "quatre"])
+    df6[3] = ["un", "deux", "trois", "quatre"]
```
Comment (ararslan) [inline test/data.jl]: lol

File: src/abstractdataframe/abstractdataframe.jl (hunk 4 above)
Comment (quinnj) [reply src/abstractdataframe/abstractdataframe.jl]: Yeah, I thought this looked fishy too.

File: src/other/utils.jl (hunk 13 above)
Comment (quinnj) [reply src/other/utils.jl]: I think it's `countnull` to allow, e.g. `DataArray`/`NullableArray`/`CategoricalArray` to define optimized versions.

File: docs/src/lib/utilities.md (hunk 1 above)
Comment (ararslan) [reply docs/src/lib/utilities.md]: nvm looks like they do different things; the former identifies complete cases as a boolean vector and the latter is now `dropnull!`.

File: src/abstractdataframe/io.jl (hunk 6 above)
Comment (ararslan) [reply src/abstractdataframe/io.jl]: Found it, it's in src/abstractdataframe/show.jl, so this is fine.

File: src/other/utils.jl (hunk 13 above)
Comment (ararslan) [reply src/other/utils.jl]: You still can quite easily:
```julia
Base.count(::typeof(isnull), a::SweetArray) = # awesome implementation
```

Comment (nalimilan) [reply src/other/utils.jl]: Yeah, let's remove it in a subsequent PR.

File: test/data.jl (hunk 15 above)
Comment (nalimilan) [reply test/data.jl]: That was really a shame for all JuliaStats! ;-)

File: src/dataframe/dataframe.jl (hunk 11 above)
Comment (nalimilan) [reply src/dataframe/dataframe.jl]: This was `Array` before the fork, and has been changed to `Vector` in DataFrames since then. I've changed the merge commit to use `Vector` since that's slightly better.

File: src/abstractdataframe/abstractdataframe.jl (hunk 4 above)
Comment (nalimilan) [reply src/abstractdataframe/abstractdataframe.jl]: Yeah, that's also related to issues around how to define `==` for `Nullable`. We should fix this in later PRs.

File: src/dataframerow/dataframerow.jl [hunk 16]
```diff
@@ -37,49 +37,76 @@ Base.done(r::DataFrameRow, s) = s > length(r)
...
-        end
+    return true
+end
+
+# lexicographic ordering on DataFrame rows, null > !null
+function Base.isless(r1::DataFrameRow, r2::DataFrameRow)
+    (ncol(r1.df) == ncol(r2.df)) ||
+        throw(ArgumentError("Rows of the data tables that have different number of columns cannot be compared ($(ncol(df1)) and $(ncol(df2)))"))
+    @inbounds for i in 1:ncol(r1.df)
+        isless(r1.df[i][r1.row], r2.df[i][r2.row]) && return true
```
Comment (nalimilan) [inline src/dataframerow/dataframerow.jl]: Since we're doing a general review, we should have a look at the semantics of this function at some point. It was added in https://github.com/JuliaData/DataTables.jl/pull/17 and I'm not sure what's its purpose. Its behavior looks quite counter-intuitive to me.

For now I've kept it, but simplified it a bit since `null` uses the right `isless` definition.

**Comment by nalimilan** [comment]
I've fixed the issues which are related to the merge. Let's handle the others later (using Femtocleaner where possible). I've also noticed a few remaining uses of `unsafe_get`, I've removed them from @quinnj's commit to keep history clean.

I've just pushed it to master since I'm afraid of what GitHub is going to do in this complex situation. We should have a look at the failures on Julia 0.7, they don't seem too hard to fix.

**Comment by quinnj** [comment]
Note that one of the 0.7 failures is a bug in current isbits Union array copying. A fix will be incoming to Base in the next few days.

File: src/dataframerow/dataframerow.jl (hunk 16 above)
Comment (nalimilan) [reply src/dataframerow/dataframerow.jl]: See https://github.com/JuliaStats/DataFrames.jl/pull/1222. Actually, the way the code was written looked weird to me, but the behavior was OK.

File: test/data.jl (hunk 15 above)
Comment (ararslan) [reply test/data.jl]: C'est dommage!
